### PR TITLE
Feature/user viewmodel

### PR DESCRIPTION
**Pull request by RagdollGoDown**
# User Viewmodel
## Description
- [x] What has been changed?
We now have a working view model for a user. It is possible to set different values like the liked recipes and ingredients in it's fridge. The view model focuses on the current user that is signed in. It will get the userId of the authenticated user. This was done, because for now the only user we are showing and modifying is the current user. This is also why on init it will check if someone is signed in.
```kotlin
init {
    userRepository.init { getCurrentUser() }
  }

  fun getCurrentUser() {
    val userId: String = firebaseAuth.currentUser?.uid ?: return
```
For now we the functionality is limited to simple operations on the user data:
- changing the username and profile picture
- adding and removing from the fridge, saved recipes and liked recipes

### Keeping the database up to date
For now, we have determined that we don't need to keep the database update at every decision from the user. The solution chosen would be to save periodically, for example every 30 seconds and immediately once the app shuts down. This is to keep the database scalable.

### Change in User class
The ids for recipes and ingredients should be in string format to simplify API calls. Their is no gain in keeping them in Long format as we use Firestore, which generates string ids.

- [x] Why was this change made?
This allows work on the UI part of the project related to the user.

When working with it, even though it updates the server periodically, every change is kept locally, so it shouldn't change anything for the UI.

## Checklist
- [x] Tests added.
- [x] Documentation updated.
- [x] All CI checks passed.
- [x] Linked issue/feature (if applicable): #88

### CODE REVIEW

**Review by AdriB02 — COMMENTED** [review]
The provided code for the `UserViewModel`, `User` and `UserRepository` is well-structured and effectively implements the management of user-related data and operations. 
However, introducing a factory pattern could streamline the creation of User, Ingredient, and Recipe objects, enhancing code maintainability and reducing duplication. 
Additionally, including more documentation inside your code would improve clarity for future reviewers.
Overall, the implementation is solid, with these enhancements offering opportunities for further improvement.

File: app/src/main/java/com/android/sample/model/user/UserViewModel.kt [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+   * Removes a recipe from the user created recipes.
+   *
+   * @param recipe the recipe to be removed
+   */
+  fun removeRecipeFromUserCreatedRecipes(recipe: Recipe) {
+    val currentCreatedRecipes = _createdRecipes.value ?: emptyList()
+    val newCreatedRecipes = currentCreatedRecipes.toMutableList()
+    newCreatedRecipes.remove(recipe)
+    _createdRecipes.value = newCreatedRecipes
+  }
```
[the remark this answers is not in the document]
Comment (RagdollGoDown) [reply app/src/main/java/com/android/sample/model/user/UserViewModel.kt:175]: Good idea

File: app/src/test/java/com/android/sample/model/user/UserViewModelTest.kt [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+    val onFailureCaptor: ArgumentCaptor<Function1<Exception, Unit>> =
+        ArgumentCaptor.forClass(clazz2)
+
+    doNothing().`when`(mockUserRepository).getUserById(any(), any(), capture(onFailureCaptor))
+
+    userViewModel.getCurrentUser()
+
+    onFailureCaptor.value.invoke(Exception())
+
```
[the remark this answers is not in the document]
Comment (RagdollGoDown) [reply app/src/test/java/com/android/sample/model/user/UserViewModelTest.kt:106]: very true

**Review by anaselmadhi — COMMENTED** [review]
# PR Review: UserViewModel Implementation

## ✅ What Looks Good

Well-structured UserViewModel with clear StateFlow usage for reactive data handling.
Comprehensive unit tests covering key functionalities, including user retrieval, updates, and list modifications.
Good use of UserRepository interface to abstract Firebase interactions, improving testability.

## ⚠️ Areas for Improvement
Code Duplication in UserViewModel:
Several functions (addIngredientToUserFridge, addRecipeToUserSavedRecipes, etc.) share similar logic for updating lists. Consider abstracting the updateList logic to reduce code duplication and improve readability.

Error Handling:
In getCurrentUser() and updateCurrentUser(), error handling is minimal. Adding logging or error handling callbacks could provide better feedback in case of Firebase failures.
Variable Naming:
In UserTest and UserViewModelTest, using clearer variable names for mocks (e.g., mockAuth instead of mockFirebaseAuth) could improve readability.
## 💡 Suggestions
Optimize updateCurrentUser(): Instead of calling getCurrentUser() immediately after updateUser(), consider passing an updated user directly to _userName, _profilePictureUrl, etc., to minimize redundant Firebase calls.

Add Additional Unit Tests:
Include tests for error scenarios in getUserById and updateUser to validate that the appropriate failure handlers are triggered.

## 📝 Checklist
 Code is well-structured and follows Kotlin and Firebase best practices.
 Documentation is clear, with function comments in UserRepository and UserViewModel.
Consider expanding test coverage for edge cases, especially for Firebase failures.


Overall, this is a strong implementation with minor areas for improvement. Good job!

File: app/src/test/java/com/android/sample/model/user/UserViewModelTest.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (AdriB02) [reply app/src/test/java/com/android/sample/model/user/UserViewModelTest.kt:106]: I meant that I would have tested it by specifying the user instead of using any().

**Comment by AdriB02** [comment]
I really like the changes you've made! The tests are much more precise now. However, there's an issue with SonarCloud that needs to be resolved before we can merge. On line 71 of [app/src/main/java/com/android/sample/model/user/UserRepositoryFirestore.kt]:
```kotlin
task: Task<Void>,
```
SonarCloud suggests replacing the usage of `Void` with `Unit`. Could you make that adjustment?

**Review by AdriB02 — APPROVED** [review]
Thank you for implementing the proposed changes! Initializing MutableStateFlow with an empty list instead of null really improves safety and readability. The refactoring into the generic updateList helper function, as suggested, effectively reduces redundancy and simplifies the logic. I also noticed that the unused line was removed, and the factory was added, which is great.

Everything looks good, and I’m happy to approve and merge this PR. Appreciate the quick turnaround on these changes!
